Add support for colon separator of keyspace:shard

bbeaudreault opened the pull request:
@sougou @michael-berlin

sougou commented on go/vt/topo/topoproto/shard.go:
```diff
@@ -34,11 +36,16 @@ func ParseKeyspaceShard(param string) (string, string, error) {
+}
+
+func parseKeyspaceShard(param string) []string {
+	rp := regexp.MustCompile("[:/]")
```
regexp is probably overkill for this. Given that this will be a high QPS call, prefer using https://golang.org/pkg/strings/#LastIndexAny instead.

sougou commented:
We'll wait for @michael-berlin to look at this also.

bbeaudreault commented:
Thanks, I changed to LastIndexAny. I also removed the test which fails on a path with multiple slashes, because with LastIndexAny it's harder to determine that. Not sure how strict you want to be about that case. I can add it back and go a different route if you'd like

bbeaudreault commented:
@michael-berlin @sougou updated per conversation in slack

michael-berlin commented on go/vt/vtgate/vtgate.go:
```diff
@@ -1043,3 +1045,15 @@ func annotateBoundShardQueriesAsUnfriendly(queries []*vtgatepb.BoundShardQuery)
+// parseKeyspaceOptionalShard parses a "keyspace/shard" or "keyspace:shard" string
+// and extracts the parts. If a shard is not specified, it's
+// returned as empty string. We need to support : and / in vtgate because some clients
+// can't support our default of slash. Everywhere else we only support /
```
nit: Missing period in the last sentence after `/`.

michael-berlin commented on go/vt/vtgate/vtgate_test.go:
```diff
@@ -2683,3 +2683,45 @@ func TestVTGateShowMetadataTwoShards(t *testing.T) {
+		keyspace:      "ks",
+		shard:         "",
+	}, {
+		keyspaceShard: "/-80",
```
I suggest that we remove the test cases where one of the parts is empty.

IMO this is outside of the "specification" and therefore the behaviour for such an input should not be defined. But by having it in the test, we're implicitly defining it and making it part of our "specification". For example, we may consider to error on this input in the future.

Therefore, I prefer if you delete this case and the cases `ks/`, `ks:` and `:-80`.

bbeaudreault commented:
@michael-berlin fixed up, thanks

michael-berlin commented:
Thanks! :)
